apns/analysis/apns2_utils.py

```diff
@@ -0,0 +1,369 @@
+    record the difference in tuple, the first element is from desc1, the second is from desc2.
+    NOTE: if parent key has different value, the child key will not be compared.
+    """
+    if desc1 == desc2:
+        return {}
+    if not desc1:
+        # if desc1 is None, then desc2 should not have tuple value
+        if any([isinstance(v, tuple) for v in desc2.values()]):
+            raise ValueError("desc2 should not have tuple value")
+        return {k: (None, v) for k, v in desc2.items()} if desc2 else {}
+    if not desc2:
+        # if desc2 is None, then desc1 should not have tuple value
+        if any([isinstance(v, tuple) for v in desc1.values()]):
+            raise ValueError("desc1 should not have tuple value")
+        return {k: (v, None) for k, v in desc1.items()} if desc1 else {}
+    
+    # if they are not concurrently None, they should be dict
+    if not (isinstance(desc1, dict) and isinstance(desc2, dict)):
+        raise ValueError("desc1 and desc2 should be dict")
+    # make sure before comparison, the value is not tuple
+    if not all([not isinstance(v, tuple) for v in desc1.values()]):
+        raise ValueError("desc1 should not have tuple value")
+    if not all([not isinstance(v, tuple) for v in desc2.values()]):
+        raise ValueError("desc2 should not have tuple value")
+
+    diff = {}
+    for k, v in desc1.items():
+        v_ = desc2.get(k, None)
+        if isinstance(v, dict):
+            _diff = cal_dict_diff(v, v_)
+            diff.update({k: _diff}) if _diff else None
+        else:
+            if v != v_:
+                diff[k] = (v, v_)
+    for k, v in desc2.items():
+        if k not in desc1:
+            diff[k] = (None, v)
+    return diff
+
+def cal_desc_diff(desc1: dict, desc2: dict) -> dict:
+    """more specifically, calculate the difference between two APNS job
+    descriptions. Because the AtomSpecies key has value as list of dict,
+    therefore make difference for each element in list and return"""
+    keys = ["AtomSpecies", "Cell", "DFTParamSet", "CellGenerator"]
+    assert set(desc1.keys()) == set(desc2.keys()), f"desc1 and desc2 should have the same keys: {desc1.keys()} != {desc2.keys()}"
+    assert set(keys) == set(desc1.keys()), f"desc1 should have keys: {keys} != {desc1.keys()}"
+    assert set(keys) == set(desc2.keys()), f"desc2 should have keys: {keys} != {desc2.keys()}"
+    # the assertation above is to ensure the correctness of structure
+    # it must be satisfied before comparison
+    diff = {}
+    # first the AtomSpecies
+    asdiff = [cal_dict_diff(as1, as2) for as1, as2 in zip(desc1["AtomSpecies"], desc2["AtomSpecies"])]
+    diff.update({"AtomSpecies": asdiff}) if not all([not asd for asd in asdiff]) else None
+    # other keys
+    for key in keys[1:]:
+        d = cal_dict_diff(desc1[key], desc2[key])
+        diff.update({key: d}) if d else None
+    return diff
+
+import unittest
```

Comment:
Thanks, I would like to place it closely before the unittest because it is not the main part of the file, this is my consideration. And the standard that moving all imports to the top of file may have risk like function name conflict, then solving conflict will cost much time.